### PR TITLE
retdec: new port

### PR DESCRIPTION
**Pull request by catap**
#### Description

Added a new port for https://github.com/avast/retdec. The port contains also `-devel` version which is used current master.

###### Type(s)

- [ ] bugfix
- [x] enhancement
- [ ] security fix

###### Tested on

macOS 11.5.1 20G80 x86_64
Xcode 12.5.1 12E507

###### Verification 
Have you

- [x] followed our [Commit Message Guidelines](https://trac.macports.org/wiki/CommitMessages)?
- [x] squashed and [minimized your commits](https://guide.macports.org/#project.github)?
- [x] checked that there aren't other open [pull requests](https://github.com/macports/macports-ports/pulls) for the same change?
- [ ] referenced existing tickets on [Trac](https://trac.macports.org/wiki/Tickets) with full URL? 
- [x] checked your Portfile with `port lint`?
- [ ] tried existing tests with `sudo port test`?
- [x] tried a full install with `sudo port -vst install`?
- [x] tested basic functionality of all binary files?

### CODE REVIEW

**Comment by catap** [comment]
The build is failed because it required a lot of time to build. More than an hour.

File: devel/retdec/Portfile
Comment (reneeotten) [inline devel/retdec/Portfile:27]: you should use `depends_build-append` here as not to overwrite dependencies added by the PortGroups; then you should remove `cmake` here as this is added by the `cmake` PG

File: devel/retdec/Portfile
Comment (reneeotten) [inline devel/retdec/Portfile:32]: this should be written as `path:bin/dot:graphviz` so that also `graphviz-devel` can satisfy the dependency

File: devel/retdec/Portfile
Comment (reneeotten) [inline devel/retdec/Portfile:31]: is `git` really a build dependency?

File: devel/retdec/Portfile
Comment (reneeotten) [inline devel/retdec/Portfile:35]: is `openssl` not needed at runtime - in other words, does anything link to the openssl library?

File: devel/retdec/Portfile
Comment (reneeotten) [inline devel/retdec/Portfile:60]: why do you need to manually blacklist compilers, ideally specifying the correct `cxx_standard` should take care of this. What errors do you get that made this necessary?

**Review by reneeotten — CHANGES_REQUESTED** [review]

File: devel/retdec/Portfile
Comment (catap) [reply devel/retdec/Portfile:32]: Thanks!

File: devel/retdec/Portfile
Comment (catap) [reply devel/retdec/Portfile:27]: Thanks!

File: devel/retdec/Portfile
Comment (catap) [reply devel/retdec/Portfile:31]: Yes, it is. `cmake` uses `git` to download some dependencies from github.

Comment (reneeotten) [reply devel/retdec/Portfile:31]: that might warrant a closer look: fetching things should ideally only happen in the "fetch" phase. I guess here it is happening in the "build" phase which is not recommended - can these dependencies be added to MacPorts and used instead of downloading them during the build? I'll check the logs later once I am behing a computer

Comment (catap) [reply devel/retdec/Portfile:31]: @reneeotten it is not just building a dependenices, it building specified forked version of dependicies.

For example it requires to use specified version of LLVM (quite old) with some patches to expose some API.

And it requires sources for this dependencies :)

Thus, fetching and patching dependecies is deeply integrated into cmake's build and I have no idea how to simple run it on fetch stage. Ok, I have some ideas but it requires nearly to rework build system that seems like out of scope for portfile.

File: devel/retdec/Portfile
Comment (catap) [reply devel/retdec/Portfile:35]: Nope, nothing is linked. Anyway, it is needed to build at least `yara` dependency and it's binary which is used inside building is linked with `libssl`. Anyway, as soon as it build => it doesn't required.

Just a proof:
```
otool -L /opt/local/bin/retdec-*                                                                                                                 (git)-[+-retdec]
/opt/local/bin/retdec-ar-extractor:
	/opt/local/lib/libz.1.dylib (compatibility version 1.0.0, current version 1.2.11)
	/usr/lib/libc++.1.dylib (compatibility version 1.0.0, current version 905.6.0)
	/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1292.100.5)
/opt/local/bin/retdec-archive-decompiler.py: is not an object file
/opt/local/bin/retdec-bin2llvmir:
	/opt/local/lib/libz.1.dylib (compatibility version 1.0.0, current version 1.2.11)
	/usr/lib/libc++.1.dylib (compatibility version 1.0.0, current version 905.6.0)
	/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1292.100.5)
/opt/local/bin/retdec-bin2pat:
	/opt/local/lib/libz.1.dylib (compatibility version 1.0.0, current version 1.2.11)
	/usr/lib/libc++.1.dylib (compatibility version 1.0.0, current version 905.6.0)
	/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1292.100.5)
/opt/local/bin/retdec-config:
	/usr/lib/libc++.1.dylib (compatibility version 1.0.0, current version 905.6.0)
	/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1292.100.5)
/opt/local/bin/retdec-config.py: is not an object file
/opt/local/bin/retdec-decompiler.py: is not an object file
/opt/local/bin/retdec-fileinfo:
	/opt/local/lib/libz.1.dylib (compatibility version 1.0.0, current version 1.2.11)
	/usr/lib/libc++.1.dylib (compatibility version 1.0.0, current version 905.6.0)
	/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1292.100.5)
/opt/local/bin/retdec-fileinfo.py: is not an object file
/opt/local/bin/retdec-getsig:
	/opt/local/lib/libz.1.dylib (compatibility version 1.0.0, current version 1.2.11)
	/usr/lib/libc++.1.dylib (compatibility version 1.0.0, current version 905.6.0)
	/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1292.100.5)
/opt/local/bin/retdec-idb2pat:
	/usr/lib/libc++.1.dylib (compatibility version 1.0.0, current version 905.6.0)
	/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1292.100.5)
/opt/local/bin/retdec-llvmir2hll:
	/opt/local/lib/libz.1.dylib (compatibility version 1.0.0, current version 1.2.11)
	/usr/lib/libc++.1.dylib (compatibility version 1.0.0, current version 905.6.0)
	/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1292.100.5)
/opt/local/bin/retdec-macho-extractor:
	/opt/local/lib/libz.1.dylib (compatibility version 1.0.0, current version 1.2.11)
	/usr/lib/libc++.1.dylib (compatibility version 1.0.0, current version 905.6.0)
	/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1292.100.5)
/opt/local/bin/retdec-pat2yara:
	/opt/local/lib/libz.1.dylib (compatibility version 1.0.0, current version 1.2.11)
	/usr/lib/libc++.1.dylib (compatibility version 1.0.0, current version 905.6.0)
	/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1292.100.5)
/opt/local/bin/retdec-signature-from-library-creator.py: is not an object file
/opt/local/bin/retdec-stacofin:
	/opt/local/lib/libz.1.dylib (compatibility version 1.0.0, current version 1.2.11)
	/usr/lib/libc++.1.dylib (compatibility version 1.0.0, current version 905.6.0)
	/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1292.100.5)
/opt/local/bin/retdec-unpacker:
	/opt/local/lib/libz.1.dylib (compatibility version 1.0.0, current version 1.2.11)
	/usr/lib/libc++.1.dylib (compatibility version 1.0.0, current version 905.6.0)
	/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1292.100.5)
/opt/local/bin/retdec-unpacker.py: is not an object file
/opt/local/bin/retdec-utils.py: is not an object file
/opt/local/bin/retdec-yarac:
	/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1292.100.5)
```

File: devel/retdec/Portfile
Comment (catap) [reply devel/retdec/Portfile]: Just setup `cxx_standard` wasn't enough or old macOS. Anyway, using of `port:macports-libcxx` more than enough => I removed blacklisted compillers.

Thanks!

File: devel/retdec/Portfile
Comment (reneeotten) [reply devel/retdec/Portfile:36]: from the logs it appears that also openssl is downloaded as a first step, so I'd conjecture that the dependency isn't actually needed here...

**Comment by reneeotten** [comment]
currently the port doesn't build, so that needs to be resolved for sure. Please check the CI logs to try and figure out what the problem is.

Other than that, the way this port  downloads its own dependencies during the build phase is a bit unconventional... I would certainly want another pair of eyes on this before possibly merging the PR. Perhaps @kencu   @cjones051073 @ryandesign want to chime in?

File: devel/retdec/Portfile
Comment (reneeotten) [inline devel/retdec/Portfile:91]: I would personally not complicate the port with variants for Python. Just pick the latest stable version (currently 3.9) and use that unconditionally.

File: devel/retdec/Portfile
Comment (reneeotten) [inline devel/retdec/Portfile:31]: you do add the dependencies for building the documentation. It does appear though to do so you wiuld need to set: 

-DRETDEC_DOC=ON to build with API documentation (requires Doxygen and Graphviz, disabled by default)

File: devel/retdec/Portfile
Comment (catap) [reply devel/retdec/Portfile]: Yes, it is much better.

File: devel/retdec/Portfile
Comment (catap) [reply devel/retdec/Portfile:31]: I can achive to generating control-flow graphs only when I enabled doxygen and graphviz ;)

I never thought that API documentation is nessesary for this tools to be hones.

**Comment by catap** [comment]
@reneeotten and port doesn't build because it can't downloads dependency on azure, as far as I can see. Anyway, I've tried to fresh installed `10.14` and both ports builded sucessfuly.

File: devel/retdec/Portfile
Comment (reneeotten) [reply devel/retdec/Portfile:31]: I didn't look carefully, only glanced over upstreams README. I assumed that you added the `doxygen` and `graphviz` dependencies to build the documentation, I don't see another reason to have them as a build dependency. If you need them for anything else at runtime, you should declare the appropriately.

Comment (catap) [reply devel/retdec/Portfile:31]: @reneeotten I need them as build dependency because if I build it without them => a part to generate control-flow graphs is escaped. I guess it is related to `graphviz` only and `doxygen` can be excluded. Let me test.

**Comment by catap** [comment]
@reneeotten and right now azure is failed with error:

```
Creating port index in /Users/runner/work/1/s/ports
Adding port devel/retdec
Adding subport retdec-devel
Failed to parse file multimedia/x265/Portfile: can't read "universal_possible": no such variable
```

**Comment by reneeotten** [comment]
> @reneeotten and right now azure is failed with error:
> 
> ```
> Creating port index in /Users/runner/work/1/s/ports
> Adding port devel/retdec
> Adding subport retdec-devel
> Failed to parse file multimedia/x265/Portfile: can't read "universal_possible": no such variable
> ```

Yeah, ignore that - Azure still doesn't use the latest MacPorts version but still v1.6.x and the `universal_possible` I think was introduced in v1.7.x.

File: devel/retdec/Portfile
Comment (catap) [reply devel/retdec/Portfile:36]: @reneeotten for example it requires in the first lines of this cmake file: https://github.com/avast/retdec/blob/master/src/fileformat/CMakeLists.txt

File: devel/retdec/Portfile
Comment (catap) [reply devel/retdec/Portfile]: You ware right => `doxygen` and `graphiz` aren't build dependency. In the fact that it needs only `graphviz` and to run.

**Comment by kencu** [comment]
This looks like a desirable port to have in MacPorts. Good job getting it sorted out.

With the way it uses so many internal deps, it will never fil our "usual" model, but perhaps that is OK. It will be somewhat harder to repair older systems, etc, but I think that's OK too. 

There will be lots of quibbles about smaller things that get worked out, eg git is a bin path dep, but those can all get tweaked as well.

If it builds and runs on the last few systems, and installs things where they belong, and generally actually works, we can likely beautify it as much as possible once it is in.

**Comment by catap** [comment]
@kencu thanks!

Yes, it is working very well on Big Sur (my laptop), and I used `-devel` port to get this decompiled output https://gist.github.com/catap/075abf613c7946c8ec972fa123bf8b7e from a binary which is produce by scala-native ;)

**Comment by reneeotten** [comment]
@catap I am trying to build this locally now; there are still some changes needed I think. Are you okay, with me pushing one additional commit with the things that I need are still needed to your branch? If so, I'll do that tonight and you can take a look - if you agree and everything looks good we can squash the commits and merge.

**Comment by catap** [comment]
@reneeotten I'm more than ok to your help with this port because it consumed much more time than I've expected to invest to it ;) Thus, I've just incited you and now you should has write access to original branch that was used to create this PR.

Parallel building is working fine but it consumes *huge* amount of memory, much more than GitHub is limited for CI. Without it CI is fails. I feel that it may create an issue with prebuidled version.

I suggest to remove it before merge to master.

**Comment by reneeotten** [comment]
@catap I have pushed a commit with some additional changes (listed in the commit message). One remaining annoying issue is that this building of many dependencies do not all [UsingTheRightCompiler](https://trac.macports.org/wiki/UsingTheRightCompiler). I fixed one of them, but it now fails with the next one (`yaramod`). You can follow the link I gave above to set-up your local installation such that it will spot these/complain about such occurrences.

File: devel/retdec/Portfile
Comment (catap) [inline devel/retdec/Portfile:34]: why? I have successfully build it on VM without xcode :)

**Comment by catap** [comment]
@reneeotten which OS have you used to test it and active fails?

File: devel/retdec/Portfile
Comment (reneeotten) [reply devel/retdec/Portfile:34]: I didn't add this, it was already in your commit - but it's certainly a good question why this would be needed 😉

**Comment by reneeotten** [comment]
> @reneeotten which OS have you used to test it and active fails?

I've tested this on 10.15.7 ; but it isn't OS dependent - if you follow the instructions for testing in the link I gave about UsingTheRightCompiler you should be able to reproduce this anywhere.

File: devel/retdec/Portfile
Comment (catap) [reply devel/retdec/Portfile:34]: Indeed!

**Comment by catap** [comment]
@reneeotten thanks, I haven't no about this elegant way to track wrong configure settings! ;)

**Comment by reneeotten** [comment]
@catap sorry, I got distracted... have you figured out how to make sure that the correct compiler is used for everything? If so, I can give this another go locally myself and then it should hopefully be ready-to-go...

**Comment by catap** [comment]
@reneeotten yep, I did. Right bow it should work fine.

**Comment by reneeotten** [comment]
> @reneeotten yep, I did. Right bow it should work fine.

I just tried this locally and indeed it now is UsingTheRightCompiler - great! However, I see some weirdness in where files are installed... A lot of things are installed correctly in the `${prefix}`, but there are also files in `/opt/local/var/macports/build/_Users_renee_Software_macports-ports_devel_retdec/retdec/work/destroot/opt/local/share/retdec/support/x86/ords/wmpencen.ord`.  That doesn't seem right... can you check whether that happens for you as well `port contents retdec`. I suspect it might have to do with the `reinplaces` in the Portfile, but would need to look at that a bit closer.

**Comment by catap** [comment]
@reneeotten very good catch!

I've solved it. Anyway, it required some time figure it out.

**Comment by reneeotten** [comment]
> @reneeotten very good catch!
> 
> I've solved it. Anyway, it required some time figure it out.

Excellent - it looks like all files are now indeed installed in proper locations. Anyway, this PR has been open long enough for people to comment on it, which didn't really happen - so I guess nobody has any further comments ;) Let's merge this and if something comes up, that can be taken care of afterwards. Thanks @catap !

**Review by reneeotten — APPROVED** [review]

**Pull request merged**
